### PR TITLE
Add rolling-window for Redis rate limiting.

### PR DESCRIPTION
**Pull request by chowchow316**
This is a prototype of using Sorted Set in Redis to implement rolling-window rate limiter (with prefetch), but without release function for now.

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/547)

### CODE REVIEW

File: adapter/redisquota/redisquota.go [hunk 1]
```diff
@@ -130,41 +140,105 @@ func (rq *redisQuota) alloc(args adapter.QuotaArgs, bestEffort bool) (adapter.Qu
...
+		// For rate-limiting use only, and no release for now.
+		if rq.rateLimit == "rolling-window" {
+			// Subsequent commands between MULTI and EXEC will be queued for atomic execution.
+			// The queued commands will not produce any real response until EXEC.
+			conn.pipeAppend("MULTI")
+			resp, _ := conn.pipeResponse()
+
+			// Remove members in the sorted set which are out of the current window
+			now := int(currentTime.UnixNano())
+			exp := now - int(d.Expiration.Nanoseconds())
```
Comment (geeknoid) [inline adapter/redisquota/redisquota.go]: Should use time.Sub to calculate delta time.

File: adapter/redisquota/redisquota.go [hunk 2]
```diff
@@ -130,41 +140,105 @@ func (rq *redisQuota) alloc(args adapter.QuotaArgs, bestEffort bool) (adapter.Qu
...
-			rq.redisError = rq.common.Logger.Errorf("Unable to get integer response from redis: %v", err)
-			return 0, time.Time{}, 0
-		}
+		// Rolling-window algorithm
+		// For rate-limiting use only, and no release for now.
+		if rq.rateLimit == "rolling-window" {
+			// Subsequent commands between MULTI and EXEC will be queued for atomic execution.
+			// The queued commands will not produce any real response until EXEC.
+			conn.pipeAppend("MULTI")
+			resp, _ := conn.pipeResponse()
```
Comment (geeknoid) [inline adapter/redisquota/redisquota.go]: Shouldn't ignore error returns here and elsewhere

File: adapter/redisquota/redisquota.go [hunk 3]
```diff
@@ -34,6 +35,9 @@ type redisQuota struct {
 	redisPool *connPool
 
 	redisError error
+
+	// algorithm for rate-limiting, fixed-window or rolling-window
```
Comment (geeknoid) [inline adapter/redisquota/redisquota.go]: Maybe add a few lines explaining the difference between the two, pros & cons, etc?

File: adapter/redisquota/redisquota.go [hunk 4]
```diff
@@ -130,41 +140,105 @@ func (rq *redisQuota) alloc(args adapter.QuotaArgs, bestEffort bool) (adapter.Qu
...
+
+			// Remove members in the sorted set which are out of the current window
+			now := int(currentTime.UnixNano())
+			exp := now - int(d.Expiration.Nanoseconds())
+			conn.pipeAppend("ZREMRANGEBYSCORE", key, 0, strconv.Itoa(exp))
+			resp, _ = conn.pipeResponse()
+
+			// Add the current request in the format of (member, score) to the sorted set, use timestamp as score here.
+			for i := 0; i < int(result); i++ {
+				conn.pipeAppend("ZADD", key, strconv.Itoa(now), strconv.Itoa(now+i))
```
Comment (geeknoid) [inline adapter/redisquota/redisquota.go]: Can do strconv.Itoa(now) outside of the loop.

File: adapter/redisquota/redisquota.go [hunk 5]
```diff
@@ -130,41 +140,105 @@ func (rq *redisQuota) alloc(args adapter.QuotaArgs, bestEffort bool) (adapter.Qu
...
-				rq.redisError = rq.common.Logger.Errorf("Could not set expire for key.")
+			// Get the cardinality of the sorted set with key.
+			conn.pipeAppend("ZCARD", key)
+			resp, _ = conn.pipeResponse()
+
+			// Atomic execution of all the previous commands, starting from "MULTI"
+			conn.pipeAppend("EXEC")
+			// This will be the actual response to EXEC.
+			resp, err = conn.pipeResponse()
+			if err != nil {
```
Comment (geeknoid) [inline adapter/redisquota/redisquota.go]: Combine this "if" with the previous line.

File: adapter/redisquota/redisquota.go (hunk 1 above)
Comment (chowchow316) [inline adapter/redisquota/redisquota.go]: Done.

File: adapter/redisquota/redisquota.go (hunk 2 above)
Comment (chowchow316) [inline adapter/redisquota/redisquota.go]: Done. However, a lot of error handlings here could not be not covered by unit tests. And that's the main reason for the low coverage now.

File: adapter/redisquota/redisquota.go [hunk 6]
```diff
@@ -34,6 +35,11 @@ type redisQuota struct {
 	redisPool *connPool
 
 	redisError error
+
+	// algorithm for rate-limiting: either fixed-window or rolling-window.
+	// The fixed-window approach can allow 2x peak specified rate, whereas the rolling-window doesn't.
+	// Even though rolling-window is less lenient , it consumes more memory in redis.
```
Comment (geeknoid) [inline adapter/redisquota/redisquota.go]: This comment should go in the proto so it'll be visible to people trying to configure the system.

The third line should say:

"The rolling-window algorithm's additional precision comes at the cost of increased redis resource usage."

**Pull request merged**